filters/private/hexer/BaseGrid.cpp

```diff
@@ -0,0 +1,258 @@
+}
+
+// Walk one contiguous segment of the boundary
+void BaseGrid::findShape(HexId root)
+{
+    m_paths.push_back(Path(root));
+    Path& path = m_paths.back();
+
+    const Segment first(root, 0);
+    Segment cur(root, 0);
+
+    do {
+        // removes possible roots that are passed over, and sets information
+        // to be used in parentOrChild()
+        if (cur.horizontal())
+        {
+            // all hexagons with non-dense neighbors at edge 0 are possible roots.
+            if (cur.edge == 0)
+                m_possibleRoots.erase(cur.hex);
+
+            // if path is at edge 3, normalize to edge 0 of hex across edge 3
+            // so hexagons can be processed separately in parentOrChild()
+            HexId pathHex = (cur.edge == 0 ? cur.hex : edgeHex(cur.hex, 3));
+            m_hexPaths.insert({pathHex, &path});
+
+            // here we set the minimum I or J coordinate, to be used for finding
+            // parents and children from m_hexPaths
+            setMinCoord(pathHex);
+        }
+        path.addPoint(findPoint(cur));
+        const auto& [left, right] = nextSegments(cur);
+        // left.hex: the hexagon we would "walk into" moving clockwise from the
+        // current segment.
+        cur = isDense(left.hex) ? left : right;
+    } while (cur != first);
+
+    path.addPoint(findPoint(cur));
+}
+
+// Finds the possibilities for the next boundary segment, moving clockwise
+std::pair<Segment, Segment> BaseGrid::nextSegments(const Segment& s) const
+{
+    //             (example with HexGrid coordinates)
+    //    ____
+    //   /    \  <---- Current segment: edge 4 of (0,0)
+    //  / 0,0  \__v----------- Possible next segment, left: edge 3 of (1,-1)
+    //  \      /<---\----- Possible next segment, right: edge 5 of (0,0)
+    //   \____/ 1,-1 \
+    //   /    \      /
+    //  / 0,-1 \____/
+    //  \      /    \
+    //   \____/      \
+    //
+    static const int next[] { 1, 2, 3, 4, 5, 0 };
+    static const int prev[] { 5, 0, 1, 2, 3, 4 };
+
+    Segment right(s.hex, next[s.edge]);
+    Segment left(edgeHex(s.hex, right.edge), prev[s.edge]);
+    return { left, right };
+}
+
+// Determine whether a path is enclosed within another
+void BaseGrid::findParentPaths()
+{
+    for (auto& p : m_paths) {
+        // the only real difference between parentOrChild in the two grid 
+        // types is whether they look down i or j.
+        parentOrChild(p);
+
+        !p.parent() ?  m_roots.push_back(&p) : p.parent()->addChild(&p);
+    }
+    for (size_t i = 0; i < m_roots.size(); ++i) {
+        m_roots[i]->finalize(CLOCKWISE);
+    }
+}
+
+// Finds whether a path should be a root path or child path
+void BaseGrid::parentOrChild(Path& p)
+{
+    // Determines how many paths are walked through in the direction of
+    // a path's parent is always null to start
+    HexId hex = p.rootHex();
+    // get the i or j component of the hexagon, depending on which indexing
```

Comment:
Sides 0 and 3 (used to find possible roots) are +i/-i for H3Grid and -j/+j for HexGrid. So if I moved in the i direction for HexGrid it might not run into possible roots. I tried to resolve this by changing the edge numbering in ibell13/hexer@a32758c but I didn't get it working. I'm not sure if the logic I have right now for it is completely 1:1 between the grid types but I know it works.

Reply:
By possible roots I meant m_hexPaths - sorry

Reply:
I guess I don't understand your naming. It seems like it just confuses things. There is a J axis for H3Grid. Why not just use that? Or just don't reference such things at all.

Reply:
I find path hexes with reference to edge 0 and edge 3. H3Grid +j/-j is edge 1 and 4. So moving through only the i or j axis will miss hexagons that are on the edges of paths since only hexagons bordering paths at side 0/3 are in m_hexPaths. I can delete the comment if it's confusing but I feel like it is useful to explain it. Or I can try to redo the edge numbering again